MovieQuiz/Presentation/MovieQuizPresenter.swift

```diff
@@ -0,0 +1,130 @@
+import UIKit
+
+final class MovieQuizPresenter: QuestionFactoryDelegate {
+    let questionsAmount: Int = 10
+    var currentQuestionIndex: Int = 0
+    var currentQuestion: QuizQuestion?
+    var correctAnswers: Int = 0
+    weak var viewController: MovieQuizViewControllerProtocol?
+    var questionFactory: QuestionFactoryProtocol?
+    var statisticService: StatisticService?
+    
+    init(viewController: MovieQuizViewControllerProtocol) {
+        self.viewController = viewController
+
+        statisticService = StatisticServiceImplementation()
+
+        questionFactory = QuestionFactory(delegate: self, moviesLoader: MoviesLoader())
+        questionFactory?.loadData()
+        viewController.showLoadingIndicator()
+        }
+    
+    func didLoadDataFromServer() {
+        viewController?.hideLoadingIndicator()
+        questionFactory?.requestNextQuestion()
+    }
+    
+    func didFailToLoadData(with error: String) {
+        viewController?.showNetworkError(message: error)
+    }
+
+    func isLastQuestion() -> Bool {
+        currentQuestionIndex == questionsAmount - 1
+    }
+    
+    func restartGame() {
+        currentQuestionIndex = 0
+        correctAnswers = 0
+        viewController?.enableButtons()
+    }
+    
+    func switchToNextQuestion() {
+        currentQuestionIndex += 1
+    }
+    
+    func didRecieveNextQuestion(question: QuizQuestion?) {
+        guard let question = question else { return }
+        currentQuestion = question
+        let viewModel = convert(model: question)
+        
+        DispatchQueue.main.async { [weak self] in
+            self?.viewController?.showQuestion(quiz: viewModel)
+        }
+    }
+    
+    func convert(model: QuizQuestion) -> QuizStepViewModel {
+        QuizStepViewModel(
+            image: UIImage(data: model.image) ?? UIImage(),
+            question: model.text,
+            questionNumber: "\(currentQuestionIndex + 1)/\(questionsAmount)"
+        )
+    }
+    
+    private func didAnswer(givenAnswer: Bool) {
+        guard let currentQuestion = currentQuestion else { return }
+        let isCorrect = givenAnswer == currentQuestion.correctAnswer
+        proceedWithAnswer(isCorrect: isCorrect)
+    }
+    
+    func didAnswer(isCorrectAnswer: Bool) {
+        if isCorrectAnswer {
+            correctAnswers += 1
+        }
+    }
+    
+    func yesButtonClicked() {
+        didAnswer(givenAnswer: true)
+    }
+        
+    func noButtonClicked() {
+        didAnswer(givenAnswer: false)
+    }
+    
+    private func proceedWithAnswer(isCorrect: Bool) {
+        didAnswer(isCorrectAnswer: isCorrect)
+
+        viewController?.highlightImageBorder(isCorrectAnswer: isCorrect)
+
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
+            guard let self = self else { return }
+            self.proceedToNextQuestionOrResults()
+        }
+    }
+    
+    func proceedToNextQuestionOrResults() {
+        if self.isLastQuestion() {
```

Comment:
если ситуация позволяет, то self принято опускать
```suggestion
        if isLastQuestion() {
```